### PR TITLE
feat(vscode):  one undo-redo step for one chat stream

### PR DESCRIPTION
**Pull request by Sma1lboy**
This PR addresses #2830.

This solves the issue where a chat stream can be undone with a single undo-redo operation.

btw, I'm thinking if I should adding undoStop after resolveEdit

### CODE REVIEW

File: clients/tabby-agent/src/lsp/ChatEditProvider.ts [hunk 1]
```diff
@@ -249,121 +256,129 @@ export class ChatEditProvider {
...
+      this.mutexAbortController = null;
+    }
+  }
+
+  private async applyWorkspaceEdit(params: ApplyWorkspaceEditParams): Promise<boolean> {
+    try {
+      const result = await this.connection.sendRequest(ApplyWorkspaceEditRequest.type, params);
+      return result;
+    } catch (error) {
+      console.warn("Extended applyEdit failed, falling back to standard method:", error);
```
Comment (icycodes) [inline clients/tabby-agent/src/lsp/ChatEditProvider.ts:370]: Please do not use console for logging. You can find many examples using `logger`.

File: clients/tabby-agent/src/lsp/ChatEditProvider.ts [hunk 2]
```diff
@@ -249,121 +256,129 @@ export class ChatEditProvider {
...
+      const match = reg.exec(buffer);
+      if (match && match[0]) {
+        if (match[0] === responseDocumentTag[0]) {
+          return "document";
+        } else if (match[0] === responseCommentTag?.[0]) {
+          return "comment";
+        }
+      }
+      return false;
+    };
```
Comment (icycodes) [inline clients/tabby-agent/src/lsp/ChatEditProvider.ts:318]: It seems there is a bug after the refactor of `findOpenTag` and `processBuffer`. The opening tag (`<GENERATEDCODE>`) is not being skipped after it is found, instead, it is always included in the content after editing.

File: clients/tabby-agent/src/lsp/ChatEditProvider.ts [hunk 3]
```diff
@@ -249,121 +256,129 @@ export class ChatEditProvider {
...
       }
+    } finally {
+      this.currentEdit = null;
+      this.mutexAbortController = null;
+    }
+  }
+
+  private async applyWorkspaceEdit(params: ApplyWorkspaceEditParams): Promise<boolean> {
+    try {
+      const result = await this.connection.sendRequest(ApplyWorkspaceEditRequest.type, params);
```
Comment (icycodes) [inline clients/tabby-agent/src/lsp/ChatEditProvider.ts:367]: We can add a client capabilities flag to indicate if the client supports this method, rather than using a try-catch fallback. However, it is acceptable for now.   
I believe this improvement could be a separate PR after this PR is merged.

**Review by icycodes — CHANGES_REQUESTED** [review]

File: clients/tabby-agent/src/lsp/ChatEditProvider.ts (hunk 3 above)
Comment (Sma1lboy) [reply clients/tabby-agent/src/lsp/ChatEditProvider.ts:367]: sorry for late response, I'll put this PR on my calendar

File: clients/tabby-agent/src/lsp/ChatEditProvider.ts (hunk 2 above)
Comment (Sma1lboy) [reply clients/tabby-agent/src/lsp/ChatEditProvider.ts]: I'm sorry, I made a logical mistake, but I have fixed it. Sorry for the late response.

**Review by icycodes — APPROVED** [review]
